File: two_pointers/234.palindrome-linked-list/234-ver1.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function (head) {
  let arr = [];
  if (!head) return false;
  while (head) {
    arr.push(head.val);
    head = head.next;
  }
  let start = 0;
  let end = arr.length - 1;
  while (start < end) {
    if (arr[start] !== arr[end]) return false;
    start++;
    end--;
  }
  return true;
};

// it works
// next time:
// use a slow and fast pointer
// 1. to find the middle point: use slow = slow.next, fast = fast.next.next
//     slow would be the middle point
// 2. define a reverse function to reverse the second half linked list
//      use prev, next, head
// 3. compare if the first half and the reversed second half is the same

File: linkList/234.palindrome-linkedlist/234-ver1.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return True
 #         find the mid
        temp = ListNode()
        temp.next = head
        mid = temp
        while temp and temp.next:
            temp = temp.next.next
            mid = mid.next
        mid = mid.next


#       reverse the mid
        prev = None
        curr = mid
        nxt = curr
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        while head and prev:
            print(head.val, prev.val)
            if head.val != prev.val:
                return False
            head = head.next
            prev = prev.next
        # if not head and not mid: return True
        return True

# another way:
        # slow, fast, prev = head, head, None
        # #  find the mid
        # while fast and fast.next:
        #     slow, fast = slow.next, fast.next.next
        # prev, slow, prev.next = slow, slow.next, None

        # # reverse the second half:
        # # the order matters!
        # while slow:
        #     slow.next, prev, slow = prev, slow, slow.next
        # fast, slow = head, prev

        # while slow:
        #     if slow.val != fast.val: return False
        #     slow, fast = slow.next, fast.next
        # return True
